Leadspace with search

IgnacioBecerra opened the pull request:
### Related Ticket(s)

{{Provide url(s) to the related ticket(s) that this pull request addresses}}

### Description

{{Add a human-readable description / detail summary of what the PR is changing and any details around how and why}}

{{If applicable, include a screenshot indicating an example or examples of what the PR is changing in the application}}

### Changelog

**New**

- {{new thing}}

**Changed**

- {{changed thing}}

**Removed**

- {{removed thing}}

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "package: styles": Carbon Expressive -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

IgnacioBecerra commented:
This component is ready to be reviewed by designers, I am still working on the scrolling sticky feature but otherwise this is is ready for design review.

shixiedesign commented:
@skuruvil will be reviewing this component.

skuruvil commented:
@IgnacioBecerra here is my QA feedback. Please let me know if you have any questions. 
**### Mono theme**
**Max, xlg and lg Breakpoints**

The text width for heading, subheading and paragraph should be calc(8col-32px)  Max-width: 640px. Right now in the test link, I can see the paragraph has been stretched completely to the 8 column 

The search component width is been defined in the visual spec as Width: 10 cols, overhangs left and right. In the test link it seems like only 8 column width 

In the Visual spec, the Bottom horizontal rule color is been defined as Style: 1px solid $ui-03, it seems darker in the test link.

<img width="711" alt="Screenshot 2021-08-05 at 2 35 46 AM" src="https://user-images.githubusercontent.com/71765369/128254768-08639749-bdb9-4ca3-ac97-7973053853bf.png">


The bottom horizontal rule is present only if the lead space search container is the same Carbon theme as the page content


<img width="1095" alt="Screenshot 2021-08-05 at 2 51 35 AM" src="https://user-images.githubusercontent.com/71765369/128256570-3fdef3de-7ff2-46aa-96db-eb5352726a3f.png">


In the default variant (when we remove the optional paragraph) it seems like there is more space between the search and the subheading. As per visual spec, the spacer is supposed to be 48px
<img width="1056" alt="Screenshot 2021-08-05 at 2 37 39 AM" src="https://user-images.githubusercontent.com/71765369/128254987-95fe7c3d-8b80-4e1a-9987-8d62b5d3478e.png">

When I tried to remove the optional subheading from the knobs the changes were not reflecting on the page. But at the same time, I was able to remove the required heading. I think we need to swap that feature. 

The close button on the search seems to persist in the test link version. It is supposed to be revealed once the user starts to enter the keyword.

**@md breakpoint (672px)** 

The text width for heading, subheading, and paragraph should be calc(6col-32px)

In visual spec the search component width is been defined for medium breakpoint as Width: 6 cols, overhangs left and right. In the test link, it seems like both the text width and search component width are 8 column

<img width="942" alt="Screenshot 2021-08-05 at 2 42 38 AM" src="https://user-images.githubusercontent.com/71765369/128255562-78cd420a-3040-4e58-a3e8-384b317ae2ea.png">

There is more space between subheading and search

**@sm breakpoint (320px)** 

In visual spec, the search component width is been defined as  (48px height) Width: 4 col, no overhang. We may need to remove the overhang. 

The space between the search and subheading should be 32px. And the space between subheading and paragraph is been defined as 24px

<img width="820" alt="Screenshot 2021-08-05 at 2 44 02 AM" src="https://user-images.githubusercontent.com/71765369/128255723-03db2c62-1ed8-46a8-923f-215ab5b6fa27.png">

### **Adjacent themes** 
**Max, Xlg and Lg Breakpoints** 

Top padding and  space between subheading and search container should be 48px

The bottom horizontal rule is present only if the lead space search container is the same Carbon theme as the page content
<img width="1054" alt="Screenshot 2021-08-05 at 2 52 27 AM" src="https://user-images.githubusercontent.com/71765369/128256668-ed21cc08-ca2d-4780-afdc-b66dee73d4b8.png">


### **With image**

Not sure about the gradient stop applied here, I can see some accessibility issues on paragraph text.
<img width="549" alt="Screenshot 2021-08-05 at 2 46 33 AM" src="https://user-images.githubusercontent.com/71765369/128256023-bf03ff0a-9bf0-425f-95e7-522fb0078bd3.png">


**@md breakpoint (672px)** 

When the background image turns to an inline image, I think we don’t need the gradient applied to the image. And we may not need a bottom horizontal rule for medium and small break point

IgnacioBecerra commented:
@skuruvil Thanks for the feedback! I addressed the width and the spacing comments. The reason behind the big gap between subheading and the search is actually because simply deleting all the text in the paragraph doesn't get rid of the paragraph component itself, which still has its own padding. When adopters don't include it in their page, the padding you're noticing won't be there, as expected. That being said I added a quick check within Storybook to "remove"/hide the component when the paragraph knob is empty.

As for the subheading not updating, we've actually seen this problem before and we had to remove the knob for the time being. ([Slack thread](https://cognitive-app.slack.com/archives/GBG66T9NK/p1619795769013200?thread_ts=1619723263.090700&cid=GBG66T9NK)) This is due to the way those components are rendered, meaning that whatever is inputted at first will be stay rendered. If editing the subheading knob and opening it in a new page it will be updated. Due to this, I've removed the Subheading knob for the time being.

>The bottom horizontal rule is present only if the lead space search container is the same Carbon theme as the page content

I was talking to Dragos about this, but there's currently no surefire way of obtaining the theme a page is using since there are many ways of implementing it. This means that I've added a note in the README for adopters to read which states that the Horizontal Rule should only be used when their container is the same theme as the page. This will have to be a manual check by the adopters to adhere to the design.

For showcase purposes I've made sure that when using the Adjacent Themes + the Storybook theme knob the horizontal rule should disappear when appropiate, i.e. Storybook is using the g100 theme and Adjacent theme is `g90-and-g100` (search container is g100).

>With image

As for this, I am currently using the Background media component, which was designed to move inline only during the small breakpoint. In the Leadspace with search specs the inline picture is happening in `sm-md` breakpoints, but since the Background Media component is technically still supposed to be showing as a background in the `md` breakpoint, the gradient is still being applied to the inline image in Leadspace with search. 

Due to this, I believe it'd be better to only make the image show inline on the `sm` breakpoint and possibly target this `md` image thing in a future issue. 

Also on a last note, after talking to Shixie we agreed the scroll functionality on the specs will be added on a separate PR in the future due to its complexity.

Let me know if you have any questions or more feedback, thanks!!

kennylam commented on packages/styles/scss/components/leadspace-with-search/leadspace-with-search.scss:
```diff
@@ -85,6 +85,212 @@
+      }
+    }
+
+    ::slotted(#{$dds-prefix}-content-block-copy) {
+      font-size: 1rem;
+    }
```
Is this being used (if so, should use token)?

kennylam commented on packages/web-components/src/components/leadspace-with-search/__stories__/README.stories.mdx:
````diff
@@ -0,0 +1,141 @@
+</dds-leadspace-with-search>
+```
+
+## Horizontal Rule Use
````
```suggestion
## Horizontal rule use
```

kennylam commented on packages/web-components/src/components/leadspace-with-search/__stories__/README.stories.mdx:
````diff
@@ -0,0 +1,141 @@
+```
+
+## Horizontal Rule Use
+The Horizontal Rule should **only** be used when the Leadspace will have the same Carbon theme as the page content,
+*OR* when using Adjacent Themes, the search container will need to be the same Carbon theme as the page content.
+
+In other words, the horizontal rule should only be included as a slotted element if, for example, the page content has
+the Gray 100 theme *and* the Leadspace With Search **also** uses the the Gray 100.
+
+When using Adjacent Themes, the theme applied **must** be 'g90-and-g100', as the latter theme must match the page content theme.
````
```suggestion
The horizontal rule should **only** be used when the leadspace will have the same Carbon theme as the page content,
*OR* when using adjacent themes, the search container will need to be the same Carbon theme as the page content.

In other words, the horizontal rule should only be included as a slotted element if, for example, the page content has
the `gray-100` theme *and* `Leadspace with search` **also** uses `gray-100`.

When using adjacent themes, the theme applied **must** be `g90-and-g100`, as the latter theme must match the page content theme.
```

kennylam commented on packages/web-components/src/components/leadspace-with-search/__stories__/leadspace-with-search.stories.ts:
```diff
@@ -0,0 +1,104 @@
+ensuring software is always ready for deployment.`;
+
+const adjacentThemes = {
+  'White/ Gray 10': ADJACENT_THEMES.WHITE_AND_G10,
+  'Gray 10/ White': ADJACENT_THEMES.G10_AND_WHITE,
+  'Gray 90/ Gray 100': ADJACENT_THEMES.G90_AND_G100,
+  'Gray 100/Gray 90': ADJACENT_THEMES.G100_AND_G90,
```
Slightly inconsistent spacing in keys here.

kennylam reviewed (commented):
Looks good @IgnacioBecerra! A few small comments. Also, `leadspace-with-search.scss` should probably be renamed to `_leadspace-with-search.scss` so it can be used as a partial, no? Looks like it needs to be added to `ibm-dotcom-styles.scss` as well.

skuruvil commented:
@IgnacioBecerra Thanks for the clarification and all the changes made. We have some minor feedback below, hope that would not take much of your time. 

I think as a standard we are following 32pix right side padding to the text width. I think that something is missing here for this xlg – 1312px breakpoint (8 col – 32px )
.![Screenshot 2021-08-06 at 8.30.38 PM.png](https://images.zenhubusercontent.com/5f6b304e7153544ff6e746c2/fb0ed135-3939-43a5-9f07-08da7a9123af)


For large breakpoint (1056px), the paragraph text width should be stretched up to 10 col – 32px, while the heading and subhead remain the same as 8 col – 32px. Sorry, this is something I have missed in my first-round review.
![Screenshot 2021-08-06 at 8.32.40 PM.png](https://images.zenhubusercontent.com/5f6b304e7153544ff6e746c2/9623edaf-35db-496a-8c17-3952c6850f26)


**For showcase purposes I've made sure that when using the Adjacent Themes + the Storybook theme knob the horizontal rule should disappear when appropriate, i.e. Storybook is using the g100 theme and Adjacent theme is g90-and-g100 (search container is g100).**
I know this is just for the showcase purpose. As per your above example, I am getting a screen like this, I think this should have a horizontal rule at the bottom?
![Screenshot 2021-08-06 at 8.34.56 PM.png](https://images.zenhubusercontent.com/5f6b304e7153544ff6e746c2/0e6e3f48-8afe-4a1b-bd7c-cca2e69611e9)

For a medium breakpoint, the text and search box width should be 6 cols (for text 6 col – 32px), I think right now we have only 5 columns width.
![Screenshot 2021-08-06 at 8.35.30 PM.png](https://images.zenhubusercontent.com/5f6b304e7153544ff6e746c2/e515d730-17cc-4e65-9d8f-2d0cad25d2a2)

I think for small breakpoints we don’t usually have the 32px padding on right for the text. 
And the search box height for the small breakpoint is 48px,  rest of the breakpoints are 64px. Even this is something I have missed mentioning in my first round review
![Screenshot 2021-08-06 at 8.37.27 PM.png](https://images.zenhubusercontent.com/5f6b304e7153544ff6e746c2/50c23bb8-a5e7-4ae7-82ff-a3755d024f5d). 

Thank you

IgnacioBecerra commented:
@skuruvil Addressed the latest grid comments in my last commit!

skuruvil commented:
@IgnacioBecerra Yes the grid looks perfect to me now. 

I just noticed one more thing. 

With type-ahead:  I know we have used the typeahead as the reference component here, however,  it's slightly different from the design spec. It should follow the Same Carbon theme as the search component. And I think there is a hover state as well in the design spec.

<img width="877" alt="Screenshot 2021-08-13 at 6 42 29 PM" src="https://user-images.githubusercontent.com/71765369/129362159-b507ab1a-54a3-4503-b3b4-c70e424dd3d4.png">


Forr dark themes the input text color is not changing based on the theme and  the focus border color is
different from the design (only for dark themes)

<img width="897" alt="Screenshot 2021-08-13 at 6 44 10 PM" src="https://user-images.githubusercontent.com/71765369/129362412-d35e3981-1325-4e6f-9726-d4eecf129d19.png">

IgnacioBecerra commented:
@skuruvil Updated! Talked to @DragosRistici about the focus color and since we're sticking with Carbon styles, we're keeping the white outline on dark themes.

skuruvil commented:
@IgnacioBecerra Thanks for accomodating all the feedback. I don't have any other changes. @DragosRistici may have a final look before pass the design approval. Thank you

kennylam commented:
@IgnacioBecerra Don't forget to add new SCSS partials to `ibm-dotcom-styles.scss`. Thanks!

DragosRistici commented:
Hey, @IgnacioBecerra let's fix these little things:


## Typeahead in general
1. Typeahead focus outline is outside:
<img width="1026" alt="Screenshot 2021-08-18 at 14 35 40" src="https://user-images.githubusercontent.com/48327647/129893629-ede92ff7-4a3f-4cd5-adf7-6cac27817c34.png">
<img width="1023" alt="Screenshot 2021-08-18 at 14 35 19" src="https://user-images.githubusercontent.com/48327647/129893635-1158652b-50e2-41e3-87c8-258661529425.png">

2. For the white typeahead, hover token on the results list item does not seems to be $hover-ui
<img width="1037" alt="Screenshot 2021-08-18 at 15 00 29" src="https://user-images.githubusercontent.com/48327647/129894785-e7951214-8a0a-4170-a814-c9f5cb699da5.png">

3. The "x" button should be always present along with the text typed by the user. They are strongly connected. Even if the search bar is out of focus. 

https://user-images.githubusercontent.com/48327647/129897036-04bf8935-e3e6-46d2-a597-907274459ef8.mov

4. Typeahead results list not scrolling at all. Instead, the page content underneath scrolls.

https://user-images.githubusercontent.com/48327647/129897825-d885489c-8311-40d1-88d2-5c94b4d59a0f.mov




## Lead space search – Web Components

### General

1. Accessibility issue
<img width="880" alt="Screenshot 2021-08-18 at 14 54 42" src="https://user-images.githubusercontent.com/48327647/129893811-7e51b473-8c3a-4ea6-9e23-39fbad52fd0d.png">


## Default

### Desktop

1. All three type containers should adhere to the same rule: 8 cols - 32px, no overhang, max-width: 640px. 
<img width="812" alt="Screenshot 2021-08-18 at 14 47 26" src="https://user-images.githubusercontent.com/48327647/129892876-48fbfb4e-3889-48e7-a693-8d953c7f2ec6.png">

2. On xlg, the top spacing drops to 32px instead of keeping the 48px. 
<img width="784" alt="Screenshot 2021-08-18 at 15 10 46" src="https://user-images.githubusercontent.com/48327647/129895855-92cc4d97-8f05-4838-9552-4ab67d946952.png">

3. On lg, the paragraph should also respect the above rule from 1.
<img width="676" alt="Screenshot 2021-08-18 at 16 03 41" src="https://user-images.githubusercontent.com/48327647/129902878-1027a956-de33-4f14-8eaf-360117e720c7.png">

### Medium

1. All type containers should be 6cols - 32px, with a max-widht of 640px (for the big medium)
<img width="751" alt="Screenshot 2021-08-18 at 16 13 39" src="https://user-images.githubusercontent.com/48327647/129904314-ebc5a8a5-520f-454d-a4aa-0ab515c4c646.png">


## With Image

### Desktop

1. This repositioning here is pushing the gradient along with the image, creating accessibility issues. 
<img width="622" alt="Screenshot 2021-08-18 at 16 39 11" src="https://user-images.githubusercontent.com/48327647/129908624-b6615d9e-3108-4358-92f1-fd3be8c5dba0.png">

https://user-images.githubusercontent.com/48327647/129908902-6b58a30d-8b99-4c64-b1e2-672f335abae8.mov

### Medium breakpoint. This one is tricky. If we go with the old approach, then the background image becomes an inline image below the search, but its 100% width, and also does not have the gradient applied (like the left side). If we go with the new approach of letting the background image inside the medium breakpoint, too, then its like in the right side image.
<img width="834" alt="Screenshot 2021-08-18 at 17 18 18" src="https://user-images.githubusercontent.com/48327647/129914924-187a10cb-1132-4bb7-a6b0-04e99943788b.png">

Right now, its a hybrid: there's the background becoming inline, but the gradient remains although there's no elements on the image to require the gradient, and its also not 100% width. 
<img width="740" alt="Screenshot 2021-08-18 at 17 21 30" src="https://user-images.githubusercontent.com/48327647/129915438-e4fa666b-2be8-4ea0-9cb9-6af006b858fb.png">

I would suggest to go with the new approach. Let me know if you have any questions.

DragosRistici commented:
Also, the React Wrapper seems to have only the `Default` story with no knobs. Let me know if this will remain or if you will update it to have knows and the `With image` story also.

IgnacioBecerra commented:
@DragosRistici  feedback addressed, but fyi the React wrapper story can't include text knobs due to the way the components are built, but I was able to add the adjacent themes knob on it.

jeffchew commented:
@DragosRistici just checking if you can take another look and if this looks good to go. Thanks!

jeffchew reviewed (approved):
LGTM!

One minor thing is we can probably get rid of the `hello world` in the codesandbox CDN example.

IgnacioBecerra commented:
@shixiedesign Addressed the feedback about focus! About adding the subheading knob, the way it's being rendered makes it so we're unable to change the text using knobs (similar to [this](https://cognitive-app.slack.com/archives/GBG66T9NK/p1619795769013200?thread_ts=1619723263.090700&cid=GBG66T9NK)).

shixiedesign commented:
Visually looks good! Two questions...

1. Clicking on a typeahead option currently doesn't do anything, not search, and not updating the input field too. Functionality wise it might be an issue? 
![Aug-25-2021 07-35-32](https://user-images.githubusercontent.com/15144993/130783863-6833d2dd-22e4-480e-8aa9-227185e04992.gif)

2. I'm not sure about Storybook's accessibility check. Do we need to worry about this? 
![image](https://user-images.githubusercontent.com/15144993/130783492-742b807b-a6e0-4e20-9e91-3f42118b7c10.png)
